### DISCUSSION TITLE
Is there only one service container for each Symfony app?

### DISCUSSION THREAD

**Discussion by nbro** [symfony/symfony#54929] [Q&A]
I'm learning about Symfony, so about services and containers. I understand that services are objects, created and maintained by the so-called service container, which we can inject as parameters to our controllers. For example, `LoggerInterface` or `EntityManagerInterfacer` are services. What I am not sure about, after having read about the topic, is if there's only one service container object for the whole Symfony application or if there are multiple ones. If multiple ones, why is that?

**Comment by kira0269** [comment]
Hi ! Welcome to the Symfony world !

You are right when you say "there's only one service container object for the whole Symfony application".
By default, with a full Symfony application, the Kernel will start with a big container which knows every service definition you've created (and from the bundles/components of Symfony).

But in fact, you can meet some "scoped" containers, which are the **service locators**. For example, if you inspect the code fomr the AbstractController from Symfony, you can see a `$container` attribute. But it is not the main container of your app. This is a smaller container exposing some useful services for your controllers. 

You can read the documentation about the service locators here [https://symfony.com/doc/current/service_container/service_subscribers_locators.html#defining-a-service-subscriber](https://symfony.com/doc/current/service_container/service_subscribers_locators.html#defining-a-service-subscriber )

Enjoy !

**Comment by ghost** [comment]
Yes, there is only one DI container that is owned by Kernel class and is instantiated when php script is being executed.

**Reply by nbro** [reply]
This is a contradiction to what the other person wrote above, that there are also scoped containers, so there can be more than 1 container.

**Reply by kira0269** [reply]
It's not totally in contradiction.
There is one DI container managed by the symfony app. 

But you can inject service locators, which are objects implementing the ContainerInterface. And in Symfony core, these service locators can be named $container (like in AbstractController).

**Reply by ghost** [reply]
btw. Service Locators are services created by the main container.
